apple/swift-docc
# Update link resolution to distinguish between failed and not-yet resolved references.

Bug/issue #, if applicable: rdar://81334652

## Summary

This update the link resolution code to make a distinction between a reference that failed to resolve and a reference that has not yet been resolved.

The main benefit of this is that each reference that fails to resolve can include information about that failure; both for local references and external references.

An extra benefit of this is that the documentation context will remember that the reference failed to resolve and won't attempt to resolve it again if the same reference appear in multiple places. In practice I don't expect this to have a meaningful performance impact.

## Dependencies

n/a

## Testing

In an existing project with documentation, for example [SlothCreator](https://developer.apple.com/documentation/xcode/slothcreator_building_docc_documentation_in_xcode):

1. Add a topic reference that will fail to resolve locally: for example `<doc:unresolvabe>`
2. Add a topic reference that will fail to resolve external: for example `<doc://com.something.external:other-unresolvabe>`

The two unresolved references should present different, more detailed, diagnostics messages.

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.
